docs/internal/expressions.md

```diff
@@ -0,0 +1,597 @@
+  let d : MyType & mut = MyType;
+  let (equals, result, error) = (d === 42);
+  // `equals` is a `Bool`, `result` is `MyType` and `error` is a `MaybeError`
+}
+```
+
+The notations used for calling functions are interchangeable and do not force semantics change (dynamic vs. static) calling, which will only be determined by being able to determine, at compile time, if the type is concrete or not.
+
+_Note: It's still under review if we'll use named parameters or forwarding arguments for calls._
+
+_Note2: Check the type document for a review on static vs. dynamic methods, constructors and destructors._
+
+## Equality
+
+Verona has two types of equality: by address and structural.
+
+### Address equality
+
+Address equality, compared with the keyword `is`, compares the address of the expressions, ie. their locations.
+Two variables can only have equal addresses if they are aliases to each other.
+
+Address equality implies type equality.
+Variables of non-compatible types are never equal.
+
+Example:
+```ts
+// Some variable
+var a : U32 & mut = 42;
+// An alias to the same address
+let b = a;
+// A new variable with the same value (but different address)
+var c : U32 & mut = 42;
+
+let eq = (a is b); // TRUE, `b` is an alias to `a`
+let ne = (a is c); // FALSE, `c` is a new variable
+
+// Types can be alias to sub-types
+interface Foo {}
+class Bar : Foo {}
+
+let d : Bar & mut = Foo;
+let e : Foo & mut = d;
+
+let eq = (d is e); // TRUE, `e` is an alias to `d` of a compatible type
+
+// Singleton types are always equal to themselves but not other singleton types
+class None {}
+class All {}
+
+let always_eq = (None is None); // TRUE
+let never_eq = (None is All); // FALSE
+```
+
+Because singleton types (other than numeric types) have no machine representation, they always compare equal on the same types, which means all variables of the same singleton type are aliases to each other.
+
+Note that, even though literals have a machine representation for their values, the addresses that they are stored for `a` and `c` are different, and therefore they are not the _same variable_, even though they have the same value.
+
+### Structural equality
+
+Structural equality, compared with the method `==` on appropriate types, is a comparison of the values in the object that the variable points to, including numeric literal values.
+It is the responsibility of each class to implement their own structural equality method.
+
+The standard library implements them for all classes that can be structurally compared.
+User objects can be compared structurally, simply by implementing the `==` method.
+Similar operators (`!=`, `<`, `>`, etc) are also implemented in the standard library and can be used by any user class in the same way.
+
+Example:
+```ts
+// Some variable
+var a : U32 & mut = 42;
+// A new variable with different value
+let b = a + 1;
+// A new variable with the same value (but different address)
+var c : U32 & mut = 42;
```

Comment:
Error, because of the `mut`.  If we have two `U32 & imm`, they will refer to the same object if they have the same value.

Reply:
not according to the new idea that `a` and `c` could be different objects with the same value.

we can discuss this on the next meeting and make a more general decision.

I'm personally happy either way.

Reply:
`a` and `c` are different pointers.  Making them `mut` makes sense *only* if they may refer to the same object such that changing one changes the other.  If you want that, you can create an object with a `U32 & imm` field and hold a `mut` reference to it in `a` and `b`.  If they are `imm` then it doesn't matter if they refer to the same or different objects: if they refer to objects with the same value then (as an optimisation) they can refer to the same object because we don't guarantee that `U32::+(U32 & imm, U32 & imm) : U32 & imm` returns a *new* immutable `U32`, it is allowed to return a pointer to an existing `U32 & imm` and (as an optimisation) we can therefore embed the value in the object.

If you have a `U32 & mut` then you have two choices:

 - It's a boxed thing (an object that has a word field) and you need to make sure it's in the right region and so on.
 - We have special casing of integers / floats that leaks all of the way into the abstract machine, rather than being purely an implementation artefact.

Neither of these makes me happy.

Reply:
I have already replaced `mut` with `imm`